core/src/main/scala/org/apache/spark/scheduler/BlacklistTracker.scala

```diff
@@ -0,0 +1,253 @@
+      strategy.getExecutorBlacklist(executorIdToFailureStatus, atomTask, clock)
+    updateBlacklistExecutorCache(atomTask, executors)
+    executors
+  }
+
+  // The actual implementation is delegated to strategy
+  def nodeBlacklist(): Set[String] = synchronized {
+    if (isBlacklistNodeCacheValid) {
+      getNodeBlacklistFromCache
+    } else {
+      val nodes = strategy.getNodeBlacklist(executorIdToFailureStatus)
+      updateBlacklistNodeCache(nodes)
+      nodes
+    }
+  }
+}
+
+/**
+ * Hide cache details in this trait to make code clean and avoid operation mistake
+ */
+private[scheduler] trait BlacklistCache {
+
+  // local cache to minimize the the work when query blacklisted executor and node
+  private val blacklistExecutorCache = mutable.HashMap.empty[StageAndPartition, Set[String]]
+  private val blacklistNodeCache = mutable.Set.empty[String]
+
+  // The flag to mark if cache is valid, it will be set to false when executorIdToFailureStatus be
+  // updated and it will be set to true, when called executorBlacklist and nodeBlacklist.
+  private var _isBlacklistExecutorCacheValid : Boolean = false
+  private var _isBlacklistNodeCacheValid: Boolean = false
+
+  private val cacheLock = new Object()
+
+  protected def isBlacklistExecutorCacheValid : Boolean = _isBlacklistExecutorCacheValid
+  protected def isBlacklistNodeCacheValid: Boolean = _isBlacklistNodeCacheValid
+
+  protected def updateBlacklistExecutorCache(
+      atomTask: StageAndPartition,
+      blacklistExecutor: Set[String]): Unit = cacheLock.synchronized {
+    if (!_isBlacklistExecutorCacheValid) {
+      blacklistExecutorCache.clear()
+    }
+    blacklistExecutorCache.update(atomTask, blacklistExecutor)
+    _isBlacklistExecutorCacheValid = true
+  }
+
+  protected def updateBlacklistNodeCache(
+      blacklistNode: Set[String]): Unit = cacheLock.synchronized {
+    if (!_isBlacklistNodeCacheValid) {
+      blacklistNodeCache.clear()
+    }
+    blacklistNodeCache ++= blacklistNode
+    _isBlacklistNodeCacheValid = true
+  }
+
+  protected def invalidateCache(): Unit = cacheLock.synchronized {
+    _isBlacklistExecutorCacheValid = false
+    _isBlacklistNodeCacheValid = false
+  }
+
+  protected def getExecutorBlacklistFromCache(
+      atomTask: StageAndPartition): Option[Set[String]] = {
+    blacklistExecutorCache.get(atomTask)
+  }
+
+  protected def getNodeBlacklistFromCache: Set[String] = blacklistNodeCache.toSet
+}
+
+/**
+ * A class to record details of failure.
+ *
+ * @param initialTime the time when failure status be created
+ * @param host the node name which running executor on
+ * @param numFailuresPerTask all tasks failed on the executor (key is StageAndPartition, value
+ *        is the number of failures of this task)
+ */
+private[scheduler] final class FailureStatus(
+    initialTime: Long,
+    val host: String,
+    val numFailuresPerTask: mutable.HashMap[StageAndPartition, Int]) {
+
+  var updatedTime = initialTime
```

Comment:
A potential issue with the design is the huge variance possible in how long two executors/nodes remain black-listed depending on how the failures get reported since time is always updated to last failure time.
Not sure what the implication of that would be though - would be interesting to monitor in subsequent releases.